flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java

```diff
@@ -311,25 +315,32 @@ public CheckpointException getFailureCause() {
 				try (CheckpointMetadataOutputStream out = targetLocation.createMetadataOutputStream()) {
 					Checkpoints.storeCheckpointMetadata(savepoint, out);
 					finalizedLocation = out.closeAndFinalizeCheckpoint();
+				}
 
+				CompletedCheckpoint completed = new CompletedCheckpoint(
+					jobId,
+					checkpointId,
+					checkpointTimestamp,
+					System.currentTimeMillis(),
+					operatorStates,
+					masterStates,
+					props,
+					finalizedLocation);
+
+				try {
+					completedCheckpointStore.addCheckpoint(completed);
+				} catch (Throwable t) {
+					completed.discardOnFailedStoring();
```

Comment:
could there be some side effects on changing order of the operations? Now those lines are executed before `inalizedLocationFuture.thenApplyAsync((completed) ...` on L343.

What would happen if there is some exception etc?

Reply:
Good question!

Actually `completedCheckpointStore.addCheckpoint` should be called before `finalizedLocationFuture.thenApplyAsync((completed)`. The `finalizedLocationFuture.thenApplyAsync((completed)` does things like completing `onCompletionPromise`, reporting completed statistics, disposing the pending checkpoint. However if `completedCheckpointStore.addCheckpoint` fails afterwards, does this checkpoint succeeds? I don't think so. But `onCompletionPromise` has been completed in this scenario. It's inconsistent here.

So the right way here is calling `completedCheckpointStore.addCheckpoint` first, then completing `onCompletionPromise`.I was planning to do this as a follow-up issue. However since we have decided to combine the finalization and adding into `completedCheckpointStore`,

Reply:
However I do find a problem inspired by this comment, nice job!

There might be a cancellation or shutting down during the finalization. It's not well handled yet, I would update the PR later to cover this scenario.

Reply:
I assume that one (or both) of the tests that you have added in the last commit were failing before fixing the issue?

Reply:
It should be.